### PR TITLE
Do not force `sync_dist=True` on epoch end

### LINKED ISSUE #13210: Do not force `sync_dist=True` on epoch end
## Proposed refactor

https://github.com/PyTorchLightning/pytorch-lightning/blob/fb40cbce2ea4afbcf5842e9ec524f665a3faa95d/pytorch_lightning/trainer/connectors/logger_connector/result.py#L524-L530

Metrics logged inside `epoch_end` hooks are forcefully synced. This makes the flag irrelevant and causes communication overrhead that might be undesired.

### Motivation

This was added https://github.com/PyTorchLightning/pytorch-lightning/pull/7966

With the reasoning that users would likely forget to set this themselves. Currently, for extra control, using `torchmetrics.Metric`s is the suggested solution.

### Pitch

Do not force this sync and show a warning for 1 epoch suggesting setting `sync_dist=True` for such metrics.

### Additional context

Proposed by @rohitgr7 

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @awaelchli @rohitgr7 @carmocca @edward-io @ananthsub @kamil-kaczmarek @Raalsky @Blaizzy

### PR DESCRIPTION
**Pull request by krishnakalyan3**
Fixes #13210

Just FYI, with fault-tolerant training, we need to ensure sync happens because during a failure on, let's say, a spot instance, the metrics are synced and saved to the checkpoint, so during restart, metrics on rank 0 are from the accumulated ones from the previous run, and on other ranks, they are 0. So we need to make sure they are synced in further training to ensure correct calculation.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/Lightning-AI/lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 1]
```diff
@@ -522,12 +522,15 @@ def _get_cache(result_metric: _ResultMetric, on_step: bool) -> Optional[Tensor]:
             cache = result_metric._forward_cache
         elif not on_step and result_metric.meta.on_epoch:
             if result_metric._computed is None:
-                # always reduce on epoch end
-                should = result_metric.meta.sync.should
-                result_metric.meta.sync.should = True
+                if not result_metric.meta.sync.should:
+                    warning_cache.warn(
+                        f"It is recommended to use `self.log({result_metric.meta.name!r}, sync_dist=True)` when"
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/logger_connector/result.py:527]:
```suggestion
                        f"It is recommended to use `self.log({result_metric.meta.name!r}, ..., sync_dist=True)` when"
```

File: tests/tests_pytorch/trainer/logging_/test_distributed_logging.py [hunk 2]
```diff
@@ -194,3 +196,18 @@ def on_test_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") ->
...
+
+
+def test_logger_sync_dist():
+    class CustomBoringModel(BoringModel):
+        def training_epoch_end(self, *args, **kwargs):
+            super().training_epoch_end(*args, **kwargs)
+            self.log("global_rank", self.global_rank, sync_dist=False)
+
+    model = CustomBoringModel()
+    trainer = Trainer(fast_dev_run=1, accelerator="cpu", strategy="ddp", devices=2)
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/logging_/test_distributed_logging.py:208]: Would it still be printed if strategy is not ddp?

I don't think it should

File: tests/tests_pytorch/trainer/logging_/test_distributed_logging.py [hunk 3]
```diff
@@ -194,3 +196,18 @@ def on_test_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") ->
...
+
+    with pytest.warns(UserWarning, match="It is recommended to use .* sync_dist=True"):
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/logging_/test_distributed_logging.py:210]: Can the match check that the repr worked as expected?

File: tests/tests_pytorch/trainer/logging_/test_distributed_logging.py [hunk 4]
```diff
@@ -194,3 +196,18 @@ def on_test_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") ->
...
+        trainer.fit(model)
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/tests_pytorch/trainer/logging_/test_distributed_logging.py:211]: it's using `warning_cache`

File: src/pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 5]
```diff
@@ -522,12 +524,23 @@ def _get_cache(result_metric: _ResultMetric, on_step: bool) -> Optional[Tensor]:
...
         elif not on_step and result_metric.meta.on_epoch:
             if result_metric._computed is None:
-                # always reduce on epoch end
                 should = result_metric.meta.sync.should
-                result_metric.meta.sync.should = True
+                if not result_metric.meta.sync.should and distributed_available():
+                    if _fault_tolerant_training():
+                        # make sure to always sync across devices when fault-tolerant is used
+                        result_metric.meta.sync.should = True
+                    else:
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/logger_connector/result.py:532]: I think this should be
```suggestion
                    elif distributed_available():
```

File: tests/tests_pytorch/trainer/logging_/test_distributed_logging.py [hunk 6]
```diff
@@ -194,3 +206,22 @@ def on_test_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") ->
     assert trainer.logger.logs == {"fit": 1, "validate": 1, "test": 1}
     trainer.predict(model)
     assert trainer.logger.logs == {"fit": 1, "validate": 1, "test": 1, "predict": 1}
+
+
+@pytest.mark.parametrize("distributed_env", [True, False])
+def test_logger_sync_dist(distributed_env):
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/logging_/test_distributed_logging.py:212]: We have similar tests in `tests/tests_pytorch/core/test_metric_result_integration.py`. I would move this one there

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 7]
```diff
@@ -522,12 +524,23 @@ def _get_cache(result_metric: _ResultMetric, on_step: bool) -> Optional[Tensor]:
             cache = result_metric._forward_cache
         elif not on_step and result_metric.meta.on_epoch:
             if result_metric._computed is None:
-                # always reduce on epoch end
                 should = result_metric.meta.sync.should
-                result_metric.meta.sync.should = True
+                if not result_metric.meta.sync.should and distributed_available():
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/logger_connector/result.py:528]: why? we need to make sure the sync check happens in distributed env.

File: src/pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 5 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/logger_connector/result.py:532]: distributed check is required for fault tolerance too.

File: src/pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 8]
```diff
@@ -522,12 +524,23 @@ def _get_cache(result_metric: _ResultMetric, on_step: bool) -> Optional[Tensor]:
...
+                    if _fault_tolerant_training():
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/logger_connector/result.py:533]: done!

File: tests/tests_pytorch/trainer/logging_/test_distributed_logging.py (hunk 6 above)
Comment (rohitgr7) [reply tests/tests_pytorch/trainer/logging_/test_distributed_logging.py]: done!

**Comment by krishnakalyan3** [comment]
Thank you @rohitgr7. It was really nice learning from you.
Thanks for the reviews @carmocca  and @otaj
